docs/documentation/namespaces-proposal.md

```diff
@@ -0,0 +1,93 @@
+# Namespaces Proposal
+
+This document contains a proposal for a guideline on how to expand the usage of 
+php namespaces to ILIAS modules and services. The suggested approach will elaborate 
+the following aspects:
+* naming conventions
+* integration of composer and autoloading
+* handling of subdirectories and interfaces
+* migration of the current structure
+* possible problems
+
+## Guideline
+
+### Naming & Structuring
+
+* The naming and structuring of all Namespaces MUST follow the PHP Standards Recommendation PSR-4: https://www.php-fig.org/psr/psr-4/.
+* The base prefix namespace is *ILIAS* and is pointing at the subdirectory */src*. Therefore, all Modules and Services MUST be located in the directory */src*. 
+* The subdirectory */classes* for Modules and Services will not be needed anymore and MUST be removed.
+* The prefixes "*class.*" and "*il*" for ILIAS classes will not be needed anymore and MUST be removed.
+* Newly created interfaces, abstract classes and traits MUST follow the PSR Naming Conventions: https://www.php-fig.org/bylaws/psr-naming-conventions. Already existing interfaces, abstract classes and traits SHOULD be renamed.
+* Every Module and Service folder MUST have a subdirectory for every type of asset (js, css, templates, images, docs) 
+existing in this Module / Service.
+
+*Examples:*
+
+* The class *ilObjCourse* would be renamed *ObjCourse*.
+* The file *class.ilObjCourse.php* would be renamed *ObjCourse.php*.
+* The file would be moved to *src/Modules/Course/*.
+* The namespace would be *ILIAS\Modules\Course*.
+* Any JavaScript files used in the Course Module would be located at *src/Modules/Course/js/*.
+
+### Autoloading & Class Imports
+
+* The composer autoloader (located at vendor/autoload.php) MUST be used for autoloading.
+* The autoloader MUST be included in every entry point class (index.php, ilias.php, ...).
+* Include statements (i.e. "require", "require_once", "include", "include_once") MUST NOT be used anywhere outside 
+of the entry point classes.
+* Existing include statements outside of the entry point classes MUST be removed.
+* The required dependencies of a class (outside of its own namespace) MUST be imported via '*use*' statements at 
+the beginning of the class definition.
+* The '*use*' statements SHOULD always import the absolute namespace of a file.
+* Fully-qualified names to call a class MUST never be used. 
```

Comment:
Just curious, I have no objections to this rule: What is the reason for this rule? I can't extract the "why" from this text :sweat_smile:

Reply:
Mostly to get a good overview of all dependencies of a class. Also (the second part) it looks messy having fully-qualified names like '\ILIAS\Modules\Course\ObjCourse()' in the code.